File: 16-3sum-closest.go
func abs(num int) int {
    if num < 0 {
        return -num
    }
    return num
}

func threeSumClosest(nums []int, target int) int {
    sum := nums[0] + nums[1] + nums[2]
    min_diff := abs(sum - target)
    sort.Ints(nums)
    for idxa, va := range nums {
        idxb := idxa + 1
        idxc := len(nums) - 1
        for idxb < idxc {
            vb := nums[idxb]
            vc := nums[idxc]
            diff := va + vb + vc - target
            if abs(diff) < min_diff{
                min_diff = abs(diff)
                sum = va + vb + vc
            }
            if diff < 0 {
                idxb += 1
            } else {
                idxc -= 1
            }
        }
    }
    return sum
}

File: bwu/array/16-3sum-closest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        n = len(nums)
        if n < 3:
            return 0
        ret = float('inf')
        
        nums.sort()
        for i in xrange(n - 2):
            new_t = target - nums[i]
            l, r = i + 1, n - 1
            while l < r:
                if nums[l] + nums[r] == new_t:
                    return target
                    
                tmp = nums[l] + nums[r] + nums[i]
                if abs(tmp - target) < abs(ret - target):
                    ret = tmp
                    
                if nums[l] + nums[r] < new_t:
                    l += 1
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1
                else:
                    r -= 1
                    while l < r and nums[r] == nums[r + 1]:
                        r -= 1
                        
        return ret
